### PR TITLE
[BugFix]Drive/Jumpjet/Ship/Teleport locomotor did not power on when it is un-piggybacked bugfix

### PR DESCRIPTION
**Pull request by NetsuNegi**
this is just a bugfix.

### CODE REVIEW

**Comment by tyuah8** [comment]
I can't tell by looking at the source code, but **what if a Chrono Miner is lifted out of the Service Depot by a Magnetron?**  Does this bug fix address this corner case?

What about other locomotors, that can be disabled, such as the hover locomotor used by the Robot Tank?  **What if a Robot Tank is lifted out of a Service Depot by a Magnetron?**  Is the Robot Tank still functional afterwards?

-----

Here is what I did to fix the problem:

```
static bool IsOnline ( TechnoClass* This )
{
	if (  This->Deactivated || This->IsUnderEMP() )
		return false;
	if (  auto Building = abstract_cast<BuildingClass*>( This ) )
	if ( !Building->IsPowerOnline() )
		return false;
	return true;
}

static void PowerOnIf ( ILocomotion* Locomotion )
{
	//
	LocomotionClass* Locomotor = static_cast<LocomotionClass*>( Locomotion );
	FootClass* Foot = Locomotor->LinkedTo;
	bool On = IsOnline( Foot );
	//

	if ( On ) Locomotor->Power_On ();
	else      Locomotor->Power_Off();
}

////

DEFINE_HOOK( 0x4AF94D,    DriveLocomotionClass__End_Piggyback__PowerOn, 0x7 )
{
	GET( ILocomotion**, Pointer, EDX );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x54DADC,  JumpjetLocomotionClass__End_Piggyback__PowerOn, 0x5 )
{
	GET( ILocomotion**, Pointer, EDI );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x69F05D,     ShipLocomotionClass__End_Piggyback__PowerOn, 0x7 )
{
	GET( ILocomotion**, Pointer, EDX );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5 )
{
	GET( ILocomotion**, Pointer, ESI );
	PowerOnIf( *Pointer );
	return NULL;
}
```

The above should handle all corner cases; although, I don't remember if anything else is needed...

**Comment by tyuah8** [comment]
Does this fix handle corner cases such as **a Magnetron lifting a Chrono Miner out of a Service Depot while the Chrono Miner is being repaired?**  Does the Chrono Miner still work afterwards?

**Comment by NetsuNegi** [comment]
> Does this fix handle corner cases such as **a Magnetron lifting a Chrono Miner out of a Service Depot while the Chrono Miner is being repaired?** Does the Chrono Miner still work afterwards?

no, but I can fix it

**Comment by NetsuNegi** [comment]
> Does this fix handle corner cases such as **a Magnetron lifting a Chrono Miner out of a Service Depot while the Chrono Miner is being repaired?** Does the Chrono Miner still work afterwards?

no, but I'll try to fix it

**Comment by NetsuNegi** [comment]
> Does this fix handle corner cases such as **a Magnetron lifting a Chrono Miner out of a Service Depot while the Chrono Miner is being repaired?** Does the Chrono Miner still work afterwards?

Alright, idk how to fix it

**Comment by tyuah8** [comment]
The following is code I am using for my own project.

```
static bool IsOnline ( TechnoClass* This )
{
	if (  This->Deactivated || This->IsUnderEMP() )
		return false;
	if (  auto Building = abstract_cast<BuildingClass*>( This ) )
	if ( !Building->IsPowerOnline() )
		return false;
	return true;
}

static void PowerOnIf ( ILocomotion* Locomotion )
{
	//
	LocomotionClass* Locomotor = static_cast<LocomotionClass*>( Locomotion );
	FootClass* Foot = Locomotor->LinkedTo;
	bool On = IsOnline( Foot );
	//

	if ( On ) Locomotor->Power_On ();
	else      Locomotor->Power_Off();
}

////

DEFINE_HOOK( 0x4AF94D,    DriveLocomotionClass__End_Piggyback__PowerOn, 0x7 )
{
	GET( ILocomotion**, Pointer, EDX );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x54DADC,  JumpjetLocomotionClass__End_Piggyback__PowerOn, 0x5 )
{
	GET( ILocomotion**, Pointer, EDI );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x69F05D,     ShipLocomotionClass__End_Piggyback__PowerOn, 0x7 )
{
	GET( ILocomotion**, Pointer, EDX );
	PowerOnIf( *Pointer );
	return NULL;
}

DEFINE_HOOK( 0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5 )
{
	GET( ILocomotion**, Pointer, ESI );
	PowerOnIf( *Pointer );
	return NULL;
}
```

**I don't remember if there is anything else that is needed**, 
but what the above code is doing is **detecting the end of a piggyback, and restoring power to the locomotor that was piggybacked**.

-----

Here is an explanation for what the code above is doing:

A Chrono Miner uses the teleport locomotor, but switches to the drive locomotor when it is driving, and switches back when it is stationary.  **When it switches to the drive locomotor, the teleport locomotor becomes "piggybacked" onto the drive locmotor**; and when it switches back to the teleport locomotor, the "piggybacking" ends, and the teleport locomotor is un-piggybacked from the drive locomotor.

When a Magnetron lifts a target, that target switches to the jumpjet locomotor, and whatever locomotor that target was using is "piggybacked" onto the jumpjet locomotor.  When the target is dropped and hits the floor, that "piggybacked" locomotor is un-piggybacked and is restored.

**The original game code did not power on a locomotor when it is un-piggybacked, so I wrote code to do the power on, which is the code you see above.**

**Comment by NetsuNegi** [comment]
> The following is code I am using for my own project.
> 
> ```
> static bool IsOnline ( TechnoClass* This )
> {
> 	if (  This->Deactivated || This->IsUnderEMP() )
> 		return false;
> 	if (  auto Building = abstract_cast<BuildingClass*>( This ) )
> 	if ( !Building->IsPowerOnline() )
> 		return false;
> 	return true;
> }
> 
> static void PowerOnIf ( ILocomotion* Locomotion )
> {
> 	//
> 	LocomotionClass* Locomotor = static_cast<LocomotionClass*>( Locomotion );
> 	FootClass* Foot = Locomotor->LinkedTo;
> 	bool On = IsOnline( Foot );
> 	//
> 
> 	if ( On ) Locomotor->Power_On ();
> 	else      Locomotor->Power_Off();
> }
> 
> ////
> 
> DEFINE_HOOK( 0x4AF94D,    DriveLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x54DADC,  JumpjetLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, EDI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x69F05D,     ShipLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, ESI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> ```
> 
> **I don't remember if there is anything else that is needed**...
> 
> ...but what the above code is doing is **detecting the end of a piggyback, and restoring power to the locomotor that was piggybacked**.
> 
> By the way, if you do decide to use the code I provided, **I don't believe your `UnitClass_SetDestination_HarvesterFix` is necessary anymore**.
> 
> Here is an explanation for what the code above is doing:
> 
> A Chrono Miner uses the teleport locomotor, but switches to the drive locomotor when it is driving, and switches back when it is stationary. **When it switches to the drive locomotor, the teleport locomotor becomes "piggybacked" onto the drive locomotor**; and when it switches back to the teleport locomotor, the "piggybacking" ends, and the teleport locomotor is un-piggybacked from the drive locomotor.
> 
> When a Magnetron lifts a target, that target switches to the jumpjet locomotor, and whatever locomotor that target was using is "piggybacked" onto the jumpjet locomotor. When the target is dropped and hits the floor, that "piggybacked" locomotor is un-piggybacked and is restored.
> 
> **You can think of piggybacking as a form of "backing-up" where the game, essentially, backs-up the previous locomotor so that it can be restored later.**
> 
> **The original game code did not power on a locomotor when it is un-piggybacked, so I wrote code to do the power on, which is the code you see above.**

ok, I will test it soon

**Comment by NetsuNegi** [comment]
> The following is code I am using for my own project.
> 
> ```
> static bool IsOnline ( TechnoClass* This )
> {
> 	if (  This->Deactivated || This->IsUnderEMP() )
> 		return false;
> 	if (  auto Building = abstract_cast<BuildingClass*>( This ) )
> 	if ( !Building->IsPowerOnline() )
> 		return false;
> 	return true;
> }
> 
> static void PowerOnIf ( ILocomotion* Locomotion )
> {
> 	//
> 	LocomotionClass* Locomotor = static_cast<LocomotionClass*>( Locomotion );
> 	FootClass* Foot = Locomotor->LinkedTo;
> 	bool On = IsOnline( Foot );
> 	//
> 
> 	if ( On ) Locomotor->Power_On ();
> 	else      Locomotor->Power_Off();
> }
> 
> ////
> 
> DEFINE_HOOK( 0x4AF94D,    DriveLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x54DADC,  JumpjetLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, EDI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x69F05D,     ShipLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, ESI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> ```
> 
> **I don't remember if there is anything else that is needed**...
> 
> ...but what the above code is doing is **detecting the end of a piggyback, and restoring power to the locomotor that was piggybacked**.
> 
> By the way, if you do decide to use the code I provided, **I don't believe your `UnitClass_SetDestination_HarvesterFix` is necessary anymore**.
> 
> Here is an explanation for what the code above is doing:
> 
> A Chrono Miner uses the teleport locomotor, but switches to the drive locomotor when it is driving, and switches back when it is stationary. **When it switches to the drive locomotor, the teleport locomotor becomes "piggybacked" onto the drive locomotor**; and when it switches back to the teleport locomotor, the "piggybacking" ends, and the teleport locomotor is un-piggybacked from the drive locomotor.
> 
> When a Magnetron lifts a target, that target switches to the jumpjet locomotor, and whatever locomotor that target was using is "piggybacked" onto the jumpjet locomotor. When the target is dropped and hits the floor, that "piggybacked" locomotor is un-piggybacked and is restored.
> 
> **You can think of piggybacking as a form of "backing-up" where the game, essentially, backs-up the previous locomotor so that it can be restored later.**
> 
> **The original game code did not power on a locomotor when it is un-piggybacked, so I wrote code to do the power on, which is the code you see above.**

works very better than my code

**Comment by NetsuNegi** [comment]
> The following is code I am using for my own project.
> 
> ```
> static bool IsOnline ( TechnoClass* This )
> {
> 	if (  This->Deactivated || This->IsUnderEMP() )
> 		return false;
> 	if (  auto Building = abstract_cast<BuildingClass*>( This ) )
> 	if ( !Building->IsPowerOnline() )
> 		return false;
> 	return true;
> }
> 
> static void PowerOnIf ( ILocomotion* Locomotion )
> {
> 	//
> 	LocomotionClass* Locomotor = static_cast<LocomotionClass*>( Locomotion );
> 	FootClass* Foot = Locomotor->LinkedTo;
> 	bool On = IsOnline( Foot );
> 	//
> 
> 	if ( On ) Locomotor->Power_On ();
> 	else      Locomotor->Power_Off();
> }
> 
> ////
> 
> DEFINE_HOOK( 0x4AF94D,    DriveLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x54DADC,  JumpjetLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, EDI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x69F05D,     ShipLocomotionClass__End_Piggyback__PowerOn, 0x7 )
> {
> 	GET( ILocomotion**, Pointer, EDX );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> 
> DEFINE_HOOK( 0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5 )
> {
> 	GET( ILocomotion**, Pointer, ESI );
> 	PowerOnIf( *Pointer );
> 	return NULL;
> }
> ```
> 
> **I don't remember if there is anything else that is needed**...
> 
> ...but what the above code is doing is **detecting the end of a piggyback, and restoring power to the locomotor that was piggybacked**.
> 
> By the way, if you do decide to use the code I provided, **I don't believe your `UnitClass_SetDestination_HarvesterFix` is necessary anymore**.
> 
> Here is an explanation for what the code above is doing:
> 
> A Chrono Miner uses the teleport locomotor, but switches to the drive locomotor when it is driving, and switches back when it is stationary. **When it switches to the drive locomotor, the teleport locomotor becomes "piggybacked" onto the drive locomotor**; and when it switches back to the teleport locomotor, the "piggybacking" ends, and the teleport locomotor is un-piggybacked from the drive locomotor.
> 
> When a Magnetron lifts a target, that target switches to the jumpjet locomotor, and whatever locomotor that target was using is "piggybacked" onto the jumpjet locomotor. When the target is dropped and hits the floor, that "piggybacked" locomotor is un-piggybacked and is restored.
> 
> **You can think of piggybacking as a form of "backing-up" where the game, essentially, backs-up the previous locomotor so that it can be restored later.**
> 
> **The original game code did not power on a locomotor when it is un-piggybacked, so I wrote code to do the power on, which is the code you see above.**

How should I describe your bug fix and what is your signature?

**Comment by tyuah8** [comment]
My... signature?  You can just reference my username, which is **[tyuah8](https://github.com/tyuah8)**.

I explain the bug fix and why it works in my previous post.  Basically the bug fix works by powering on locomotors when "piggybacking" ends (when we switch back to the piggybacked locomotor), and is needed because the Service Depot powered off the teleport locomotor.

**Comment by NetsuNegi** [comment]
> My... signature? You can just reference my username, which is **[tyuah8](https://github.com/tyuah8)**.
> 
> I explain the bug fix and why it works in my previous post. Basically the bug fix works by powering on locomotors when "piggybacking" ends (when we switch back to the piggybacked locomotor), and is needed because the Service Depot powered off the teleport locomotor.
> 
> You are free to mention my explanation as well as the reason for the fix and that is to get the Chrono Miner to work after using the Service Depot.

ok

File: src/Misc/Hooks.BugFixes.cpp [hunk 1]
```diff
@@ -968,3 +968,45 @@ DEFINE_HOOK(0x5B11DD, MechLocomotionClass_ProcessMoving_SlowdownDistance, 0x9)
 }
 
 DEFINE_JUMP(LJMP, 0x517FF5, 0x518016); // Warhead with InfDeath=9 versus infantry in air
+
+#pragma region End_Piggyback PowerOn
+// Auther: tyuah8
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:973]:
```suggestion
#pragma region End_Piggyback PowerOn
// Author: tyuah8

```

File: src/Misc/Hooks.BugFixes.cpp [hunk 2]
```diff
@@ -968,3 +968,45 @@ DEFINE_HOOK(0x5B11DD, MechLocomotionClass_ProcessMoving_SlowdownDistance, 0x9)
...
+	return 0;
+}
+
+DEFINE_HOOK(0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5)
+{
+	GET(ILocomotion*, loco, ECX);
+	End_Piggyback_PowerOn(loco);
+	return 0;
+}
+#pragma endregion
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:1012]:
```suggestion
}

#pragma endregion
```

**Review by Metadorius — APPROVED** [review]

**Comment by Metadorius** [comment]
@NetsuNegi @tyuah8 Does this bugfix also allow using locomotors other than Jumpjet for magnetron weapon? For example Teleport locomotor in order to teleport enemy unit to oneself.

Not a requirement, just curious. Would be good to fix eventually.

**Comment by tyuah8** [comment]
> @NetsuNegi @tyuah8 Does this bugfix also allow using locomotors other than Jumpjet for magnetron weapon? For example Teleport locomotor in order to teleport enemy unit to oneself.
> 
> Not a requirement, just curious. Would be good to fix eventually.

No, this bugfix simply powers-on a locomotor when it is un-piggybacked and does not change the functionality of an `IsLocomotor=yes` warhead.

File: src/Misc/Hooks.BugFixes.cpp [hunk 3]
```diff
@@ -1079,4 +1079,44 @@ DEFINE_HOOK(0x743664, UnitClass_ReadFromINI_Follower3, 0x6)
 	return SkipGameCode;
 }
 
+#pragma region End_Piggyback PowerOn
+// Author: tyuah8
+static void End_Piggyback_PowerOn(ILocomotion* loco)
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:1084]:
```suggestion
#pragma region End_Piggyback PowerOn

// Author: tyuah8
static void End_Piggyback_PowerOn(ILocomotion* loco)
```

File: src/Misc/Hooks.BugFixes.cpp [hunk 4]
```diff
@@ -1079,4 +1079,44 @@ DEFINE_HOOK(0x743664, UnitClass_ReadFromINI_Follower3, 0x6)
...
+	return 0;
+}
+
+DEFINE_HOOK(0x719F17, TeleportLocomotionClass__End_Piggyback__PowerOn, 0x5)
+{
+	GET(ILocomotion*, loco, ECX);
+	End_Piggyback_PowerOn(loco);
+	return 0;
+}
 #pragma endregion
```
Comment (Metadorius) [inline src/Misc/Hooks.BugFixes.cpp:1122]:
```suggestion
}

#pragma endregion
```

**Pull request merged**